### PR TITLE
client-go: add a Prometheus adapter for metrics

### LINKED ISSUE #35: Add prometheus metric package
The `tools/metrics` package already exists. Prometheus as the default integration does not yet.

The following two PRs should be resolved before adding it:
* kubernetes/kubernetes#36702
* kubernetes/kubernetes#36704

### PR DESCRIPTION
**Pull request by milesbxf**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

client-go metrics are extremely useful for understanding the behaviour of applications interacting with the Kubernetes API. For example, you can alert on certificate expiry or easily see if the client-side rate limiter is slowing down requests.

Screenshot 2023-05-17 at 15 23 58

Unfortunately, client-go only exposes its own metrics abstraction, and it's currently up to the user to implement any integrations themselves. There is a [Prometheus implementation available in k8s.io/component-base](https://github.com/kubernetes/component-base/blob/18782b4b48a05c81a098f3be4c6665bdab5c851d/metrics/prometheus/restclient/metrics.go), but it uses the internal component-base metrics library and is really just intended for internal Kubernetes component use.

This PR provides a standard Prometheus implementation which users can easily integrate into their own application (with `metrics.RegisterPrometheus()`. The metric names here are standardized and correlate with the metrics used for internal Kubernetes components.

Whilst this does add a dependency on prometheus/client_golang, I think the benefits are worth it:
* Many users are likely to be using Prometheus
* This makes it easy for users to instrument client-go apps when they might not otherwise
* For users maintaining their own Prometheus metrics implementations, this provides a standard library-based implementation with standard metric names that always keeps up with client-go metrics

#### Which issue(s) this PR fixes:
Fixes https://github.com/kubernetes/client-go/issues/35
Also see https://github.com/kubernetes/client-go/issues/756

#### Does this PR introduce a user-facing change?
```release-note
Added a Prometheus adapter for client-go users
```

### CODE REVIEW

**Comment by aojea** [comment]
> Unfortunately, client-go only exposes its own metrics abstraction, and it's currently up to the user to implement any integrations themselves. There is a [Prometheus implementation available in k8s.io/component-base](https://github.com/kubernetes/component-base/blob/18782b4b48a05c81a098f3be4c6665bdab5c851d/metrics/prometheus/restclient/metrics.go), but it uses the internal component-base metrics library and is really just intended for internal Kubernetes component use.

I don't think that is true, any project can import the metrics by importing them, the `init()` functions will load them

```
_ "k8s.io/component-base/metrics/prometheus/clientgo" // for client metric registration
```

you can see that there are more projects outside of kubernetes importing the metrics 
https://grep.app/search?current=3&q=_%20%22k8s.io/component-base/metrics/prometheus/clientgo%22

/hold

**Comment by milesbxf** [comment]
> > Unfortunately, client-go only exposes its own metrics abstraction, and it's currently up to the user to implement any integrations themselves. There is a [Prometheus implementation available in k8s.io/component-base](https://github.com/kubernetes/component-base/blob/18782b4b48a05c81a098f3be4c6665bdab5c851d/metrics/prometheus/restclient/metrics.go), but it uses the internal component-base metrics library and is really just intended for internal Kubernetes component use.
> 
> I don't think that is true, any project can import the metrics by importing them, the `init()` functions will load them
> 
> ```
> _ "k8s.io/component-base/metrics/prometheus/clientgo" // for client metric registration
> ```
> 
> you can see that there are more projects outside of kubernetes importing the metrics https://grep.app/search?current=3&q=_%20%22k8s.io/component-base/metrics/prometheus/clientgo%22
> 
> /hold

I see - it wasn't clear to me at all that component-base is also intended for external use by client applications. It's also not quite clear to me how to integrate the kubeRegistry with our own (non-global) PrometheusRegistry internally, but I'll have a play around and see if I can get it working, and if so, perhaps I'll propose a docs change to make this easier for others to work out in future

**Comment by aojea** [comment]
> I see - it wasn't clear to me at all that component-base is also intended for external use by client applications. It's also not quite clear to me how to integrate the kubeRegistry with our own (non-global) PrometheusRegistry internally, but I'll have a play around and see if I can get it working, and if so, perhaps I'll propose a docs change to make this easier for others to work out in future

@milesbxf better consult with sig-instrumentation people, they will be able to guide you better, I may be made some wrong assumptions in your claim

File: staging/src/k8s.io/client-go/go.mod [hunk 1]
```diff
@@ -44,13 +47,17 @@ require (
...
 	github.com/moby/spdystream v0.2.0 // indirect
 	github.com/modern-go/concurrent v0.0.0-20180306012644-bacd9c7ef1dd // indirect
 	github.com/modern-go/reflect2 v1.0.2 // indirect
 	github.com/munnerz/goautoneg v0.0.0-20191010083416-a7dc8b61c822 // indirect
 	github.com/onsi/gomega v1.27.6 // indirect
 	github.com/pkg/errors v0.9.1 // indirect
 	github.com/pmezard/go-difflib v1.0.0 // indirect
+	github.com/prometheus/client_model v0.3.0 // indirect
+	github.com/prometheus/common v0.42.0 // indirect
+	github.com/prometheus/procfs v0.9.0 // indirect
```
Comment (logicalhan) [inline staging/src/k8s.io/client-go/go.mod:60]: The entire point of the client-go abstraction was to get rid of the prometheus dependency, we cannot do this. You can already inject prometheus using the interface that client-go describes.

Comment (milesbxf) [reply staging/src/k8s.io/client-go/go.mod:60]: > You can already inject prometheus using the interface that client-go describes.

So is your recommendation that users write their own Prometheus adapter code, or should they use `k8s.io/component-base/metrics/prometheus/clientgo` as suggested in another comment?

**Comment by dgrisonnet** [comment]
@milesbxf in response to your previous comment about how to mix the registry wrapper that we have in Kubernetes and the one from client_golang.

Here is what we do in [metrics-server](https://github.com/kubernetes-sigs/metrics-server/blob/796fc0f832c1ac444c44f88a952be87524456e07/pkg/server/config.go#L105-L120):
```go
func main() {
        metricsHandler, err := c.metricsHandler()
	if err != nil {
		return nil, err
	}
        genericServer.Handler.NonGoRestfulMux.HandleFunc("/metrics", metricsHandler)
}

func (c Config) metricsHandler() (http.HandlerFunc, error) {
	// Create registry for Metrics Server metrics
	registry := metrics.NewKubeRegistry()
	err := RegisterMetrics(registry, c.MetricResolution)
	if err != nil {
		return nil, err
	}
	// Register apiserver metrics in legacy registry
	apimetrics.Register()

	// Return handler that serves metrics from both legacy and Metrics Server registry
	return func(w http.ResponseWriter, req *http.Request) {
		legacyregistry.Handler().ServeHTTP(w, req)
		metrics.HandlerFor(registry, metrics.HandlerOpts{}).ServeHTTP(w, req)
	}, nil
}
```

The first registry that we serve is `legacyregistry`, the global registry used in the Kubernetes codebase. Once you are serving metrics for this registry, you just need to import the metrics you want from client-go, similar to https://github.com/kubernetes/kubernetes/blob/4f60a8d493ab9571eb328b9d98da477a50bc7446/staging/src/k8s.io/component-base/metrics/prometheus/clientgo/metrics.go.
The second one is a `metrics.NewKubeRegistry()` to expose metrics-server's own metrics. This one is created via the wrapper, but it could be a vanilla registry from client_golang.

**Comment by logicalhan** [comment]
> The second one is a `metrics.NewKubeRegistry()` to expose metrics-server's own metrics. This one is created via the wrapper, but it could be a vanilla registry from client_golang.

To be clear, I believe @dgrisonnet is referring to the prometheus client when he says `it could be a vanilla registry from client_golang`, *not* `client-go` from k8s, where we cannot have a direct prometheus dependency.

**Comment by milesbxf** [comment]
> @milesbxf in response to your previous comment about how to mix the registry wrapper that we have in Kubernetes and the one from client_golang.

Aha - I hadn't considered the idea of using both registries in parallel then aggregating in the metrics handler - that works for our use case. If this is the recommended way for client-go users to export metrics, would you be happy for me to add some docs to client-go to make this clearer?

> That is correct. The Kubernetes stance is as @logicalhan described before, to not have any dependency on Prometheus, but there still exist ways to have both coexist in your implementations and the example I shared above should allow you to do just that.

Using component-base resolves my use-case of being able to reuse shared code to create the metrics, so I'm happy with that. Purely out of curiosity, what's the reason for this stance? Is the idea to minimise version conflicts (e.g. if a client-go user has an older version of Prometheus client_golang they can't upgrade easily)?

**Comment by dgrisonnet** [comment]
>  If this is the recommended way for client-go users to export metrics, would you be happy for me to add some docs to client-go to make this clearer?

Definitely, the more documentation we have the better.

> Purely out of curiosity, what's the reason for this stance? Is the idea to minimise version conflicts (e.g. if a client-go user has an older version of Prometheus client_golang they can't upgrade easily)?

@logicalhan has all the context, but the main reason is to make sure that all the metrics in Kubernetes have a stability level and follow our metrics stability framework: https://kubernetes.io/blog/2021/04/23/kubernetes-release-1.21-metrics-stability-ga/. You can also look at https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/3498-extending-stability which depicts the more recent version of the stability framework.

**Comment by logicalhan** [comment]
> @logicalhan has all the context, but the main reason is to make sure that all the metrics in Kubernetes have a stability level and follow our metrics stability framework: https://kubernetes.io/blog/2021/04/23/kubernetes-release-1.21-metrics-stability-ga/. You can also look at https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/3498-extending-stability which depicts the more recent version of the stability framework.

Actually the client-go changes pre-date the stability framework. The reason client-go uses a wrapper is to avoid dragging in all of prometheus every time someone has to use client-go, since they could be using open-telemetry or some other framework.

**Pull request closed without merging**